File: RoboRIO/src/main/java/org/usfirst/frc/team449/robot/jacksonWrappers/MappedPIDController.java
package org.usfirst.frc.team449.robot.jacksonWrappers;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import edu.wpi.first.wpilibj.controller.PIDController;
import io.github.oblarg.oblog.Loggable;
import io.github.oblarg.oblog.annotations.Log;
import org.jetbrains.annotations.Nullable;

@JsonTypeInfo(
    use = JsonTypeInfo.Id.CLASS,
    include = JsonTypeInfo.As.WRAPPER_OBJECT,
    property = "@class")
public class MappedPIDController extends PIDController implements Loggable {

  private final String name;

  private double measurement;
  private double output;

  @JsonCreator
  public MappedPIDController(
      final double Kp, final double Ki, final double Kd, @Nullable String name) {
    super(Kp, Ki, Kd);

    if (name == null) {
      name = "PIDController";
    }
    this.name = name;
  }

  @Override
  @Log
  public double getVelocityError() {
    return super.getVelocityError();
  }

  @Override
  @Log
  public double getSetpoint() {
    return super.getSetpoint();
  }

  @Override
  public double calculate(final double measurement) {
    this.measurement = measurement;
    return output = super.calculate(measurement);
  }

  @Log
  public double getMeasurement() {
    return measurement;
  }

  @Log
  public double getOutput() {
    return output;
  }

  @Override
  public String configureLogName() {
    return name;
  }
}

File: RoboRIO/src/main/kotlin/org/usfirst/frc/team449/robot/jacksonWrappers/MappedPIDController.kt
package org.usfirst.frc.team449.robot.jacksonWrappers

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonTypeInfo
import edu.wpi.first.wpilibj.controller.PIDController
import io.github.oblarg.oblog.Loggable
import io.github.oblarg.oblog.annotations.Log

@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.WRAPPER_OBJECT, property = "@class")
class MappedPIDController @JsonCreator constructor(
    Kp: Double, Ki: Double, Kd: Double, name: String?
) : PIDController(Kp, Ki, Kd), Loggable {
    private val name: String

    @get:Log
    var measurement = 0.0
        private set

    @get:Log
    var output = 0.0
        private set

    @Log
    override fun getVelocityError(): Double {
        return super.getVelocityError()
    }

    @Log
    override fun getSetpoint(): Double {
        return super.getSetpoint()
    }

    override fun calculate(measurement: Double): Double {
        this.measurement = measurement
        return super.calculate(measurement).also { output = it }
    }

    override fun configureLogName(): String {
        return name
    }

    init {
        var name = name
        if (name == null) {
            name = "PIDController"
        }
        this.name = name
    }
}